File: Objective-C Version/JumpStarDemo/CoreAnimations/JumpStarView/JumpStarView.m
//
//  JumpStarView.m
//  CoreAnimations
//
//  Created by Kitten Yang on 9/17/15.
//  Copyright (c) 2015 Kitten Yang. All rights reserved.
//

#import "JumpStarView.h"


#define jumpDuration 0.125
#define downDuration 0.215

@interface JumpStarView()
@property(nonatomic,strong)UIImageView *starView;
@property(nonatomic,strong)UIImageView *shadowView;
@end

@implementation JumpStarView{
    BOOL animating;
}

-(id)init{
    self = [super init];
    if (self) {

    }
    return self;
}

-(void)layoutSubviews{
    
    [super layoutSubviews];
    self.backgroundColor = [UIColor clearColor];
    if (self.starView == nil) {
        self.starView = [[UIImageView alloc]initWithFrame:CGRectMake(self.bounds.size.width/2 - (self.bounds.size.width-6)/2, 0, self.bounds.size.width-6, self.bounds.size.height - 6)];
        self.starView.contentMode = UIViewContentModeScaleToFill;
        [self addSubview:self.starView];
    }
    if (self.shadowView == nil) {
        self.shadowView = [[UIImageView alloc]initWithFrame:CGRectMake(self.bounds.size.width/2 - 10/2, self.bounds.size.height - 3, 10, 3)];
        self.shadowView.alpha = 0.4;
        self.shadowView.image = [UIImage imageNamed:@"shadow_new"];
        [self addSubview:self.shadowView];
    }
    
}

-(void)setState:(STATE)state{
    _state = state;
    self.starView.image = _state==Mark? _markedImage : _non_markedImage;
}

//上弹动画
-(void)animate{

    if (animating == YES) {
        return;
    }
    
    animating = YES;
    CABasicAnimation *transformAnima = [CABasicAnimation animationWithKeyPath:@"transform.rotation.y"];
    transformAnima.fromValue = @(0);
    transformAnima.toValue = @(M_PI_2);
    transformAnima.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
    
    CABasicAnimation *positionAnima = [CABasicAnimation animationWithKeyPath:@"position.y"];
    positionAnima.fromValue = @(self.starView.center.y);
    positionAnima.toValue = @(self.starView.center.y - 14);
    positionAnima.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    
    
    CAAnimationGroup *animGroup = [CAAnimationGroup animation];
    animGroup.duration = jumpDuration;
    animGroup.fillMode = kCAFillModeForwards;
    animGroup.removedOnCompletion = NO;
    animGroup.delegate = self;
    animGroup.animations = @[transformAnima,positionAnima];
    
    [self.starView.layer addAnimation:animGroup forKey:@"jumpUp"];
}

- (void)animationDidStart:(CAAnimation *)anim{
    
    if ([anim isEqual:[self.starView.layer animationForKey:@"jumpUp"]]) {
        [UIView animateWithDuration:jumpDuration delay:0.0f options:UIViewAnimationOptionCurveEaseOut animations:^{
            _shadowView.alpha = 0.2;
            _shadowView.bounds = CGRectMake(0, 0, _shadowView.bounds.size.width*1.6, _shadowView.bounds.size.height);
        } completion:NULL];
        
    }else if ([anim isEqual:[self.starView.layer animationForKey:@"jumpDown"]]){
        
        [UIView animateWithDuration:jumpDuration delay:0.0f options:UIViewAnimationOptionCurveEaseOut animations:^{
            
            _shadowView.alpha = 0.4;
            _shadowView.bounds = CGRectMake(0, 0, _shadowView.bounds.size.width/1.6, _shadowView.bounds.size.height);
            
        } completion:NULL];
        
    }
}

//下落动画
- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag{
    
    
    if ([anim isEqual:[self.starView.layer animationForKey:@"jumpUp"]]) {
        
        self.state = self.state==Mark?NONMark:Mark;
        NSLog(@"state:%ld",_state);
        CABasicAnimation *transformAnima = [CABasicAnimation animationWithKeyPath:@"transform.rotation.y"];
        transformAnima.fromValue = @(M_PI_2);
        transformAnima.toValue = @(M_PI);
        transformAnima.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
        
        CABasicAnimation *positionAnima = [CABasicAnimation animationWithKeyPath:@"position.y"];
        positionAnima.fromValue = @(self.starView.center.y - 14);
        positionAnima.toValue = @(self.starView.center.y);
        positionAnima.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseIn];
        
        CAAnimationGroup *animGroup = [CAAnimationGroup animation];
        animGroup.duration = downDuration;
        animGroup.fillMode = kCAFillModeForwards;
        animGroup.removedOnCompletion = NO;
        animGroup.delegate = self;
        animGroup.animations = @[transformAnima,positionAnima];
        
        [self.starView.layer addAnimation:animGroup forKey:@"jumpDown"];
        
    }else if([anim isEqual:[self.starView.layer animationForKey:@"jumpDown"]]){
        
        [self.starView.layer removeAllAnimations];
        animating = NO;
    }
    
}

@end

File: Swift Version/JumpStarDemo-Swift/JumpStarDemo-Swift/Classes/JumpStarView.swift
//
//  JumpStarView.swift
//  JumpStarDemo-Swift
//
//  Created by Kitten Yang on 1/6/16.
//  Copyright © 2016 Kitten Yang. All rights reserved.
//

import UIKit

enum SelectState {
    case NotMarked;
    case Marked
}

struct JumpStarOptions {
    var markedImage: UIImage
    var notMarkedImage: UIImage
    var jumpDuration: NSTimeInterval
    var downDuration: NSTimeInterval
}

class JumpStarView: UIView {
    
    var option: JumpStarOptions?
    
    var state: SelectState = .NotMarked {
        didSet{
            starView.image = state == .Marked ? option?.markedImage : option?.notMarkedImage
        }
    }
    
    private var starView: UIImageView!
    private var shadowView: UIImageView!
    private var animating: Bool = false
    
    override func layoutSubviews() {
        super.layoutSubviews()
        backgroundColor = UIColor.clearColor()
        if starView == nil {
            starView = UIImageView(frame: CGRect(x: bounds.width/2 - (bounds.width-6)/2, y: 0, width: bounds.width-6, height: bounds.height-6))
            starView.contentMode = .ScaleToFill
            addSubview(starView)
        }
        
        if shadowView == nil {
            shadowView = UIImageView(frame: CGRect(x: bounds.width/2 - 10/2, y: bounds.height - 3, width: 10, height: 3))
            shadowView.alpha = 0.4
            shadowView.image = UIImage(named: "shadow_new")
            addSubview(shadowView)
        }
    }
    
    func animate() {
        if animating {
            return
        }
        
        animating = true
        let transformAnimation = CABasicAnimation(keyPath: "transform.rotation.y")
        transformAnimation.fromValue = 0.0
        transformAnimation.toValue   = M_PI_2
        transformAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
        
        let positionAnimation = CABasicAnimation(keyPath: "position.y")
        positionAnimation.fromValue = starView.center.y
        positionAnimation.toValue   = starView.center.y - 14.0
        positionAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
        
        let animationGroup = CAAnimationGroup()
        if let duration = option?.jumpDuration {
            animationGroup.duration = duration
        }
        animationGroup.fillMode = kCAFillModeForwards
        animationGroup.removedOnCompletion = false
        animationGroup.delegate = self
        animationGroup.animations = [transformAnimation, positionAnimation]
        
        starView.layer.addAnimation(animationGroup, forKey: "jumpUp")
    }
    
}

// MARK : CAAnimationDelegate

extension JumpStarView {
    override func animationDidStart(anim: CAAnimation) {
        guard let _ = option else {
            return
        }
        
        if anim == starView.layer.animationForKey("jumpUp") {
            UIView.animateWithDuration(option!.jumpDuration, delay: 0.0, options: .CurveEaseOut, animations: { () -> Void in
                self.shadowView.alpha = 0.2
                self.shadowView.bounds = CGRect(x: 0, y: 0, width: self.shadowView.bounds.width*1.6, height: self.shadowView.bounds.height)
                }, completion: nil)
        } else if anim == starView.layer.animationForKey("jumpDown") {
            UIView.animateWithDuration(option!.jumpDuration, delay: 0.0, options: .CurveEaseOut, animations: { () -> Void in
                self.shadowView.alpha = 0.4
                self.shadowView.bounds = CGRect(x: 0, y: 0, width: self.shadowView.bounds.width/1.6, height: self.shadowView.bounds.height)
                }, completion: nil)
            
        }
        
    }
    
    override func animationDidStop(anim: CAAnimation, finished flag: Bool) {
        if anim == starView.layer.animationForKey("jumpUp") {
            state = state == .Marked ? .NotMarked : .Marked
            
            let transformAnimation = CABasicAnimation(keyPath: "transform.rotation.y")
            transformAnimation.fromValue = M_PI_2
            transformAnimation.toValue = M_PI
            transformAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseInEaseOut)
            
            let positionAnimation = CABasicAnimation(keyPath: "position.y")
            positionAnimation.fromValue = starView.center.y - 14
            positionAnimation.toValue = starView.center.y
            positionAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseIn)
            
            let animationGroup = CAAnimationGroup()
            if let duration = option?.downDuration {
                animationGroup.duration = duration
            }
            animationGroup.fillMode = kCAFillModeForwards
            animationGroup.removedOnCompletion = false
            animationGroup.delegate = self
            animationGroup.animations = [transformAnimation, positionAnimation]
            
            starView.layer.addAnimation(animationGroup, forKey: "jumpDown")
        } else if anim == starView.layer.animationForKey("jumpDown") {
            starView.layer.removeAllAnimations()
            animating = false
        }
    }
    
}





